File: app/src/main/java/com/sumian/hw/improve/report/dailyreport/DailyAdapter.java
package com.sumian.hw.improve.report.dailyreport;

import android.support.annotation.NonNull;
import android.support.v4.widget.NestedScrollView;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ScrollView;
import android.widget.TextView;

import com.sumian.sleepdoctor.R;
import com.sumian.hw.improve.report.note.SleepNote;
import com.sumian.hw.improve.widget.ReportSleepDurationView;
import com.sumian.hw.improve.widget.SwitchDateView;
import com.sumian.hw.improve.widget.TouchDailySleepHistogramView;
import com.sumian.hw.improve.widget.note.SleepNoteView;
import com.sumian.hw.widget.refresh.BlueRefreshView;

import java.util.ArrayList;
import java.util.List;

import butterknife.ButterKnife;

/**
 * Created by jzz
 * on 2018/3/12.
 * desc:
 */

public class DailyAdapter extends RecyclerView.Adapter<DailyAdapter.ViewHolder> {

    private ArrayList<DailyReport> mData;
    private SwitchDateView.OnSwitchDateListener mOnSwitchDateListener;
    private View.OnClickListener mOnClickListener;
    private OnRefreshCallback mOnRefreshCallback;

    DailyAdapter() {
        this.mData = new ArrayList<>(0);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.hw_lay_item_today_repot, parent, false);
        ViewHolder viewHolder = new ViewHolder(itemView);
        viewHolder.itemView.setTag(viewHolder);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        DailyReport dailyReport = mData.get(position);
        holder.initView(dailyReport, mOnSwitchDateListener, mOnClickListener, mOnRefreshCallback);
    }

    public DailyAdapter setOnSwitchDateListener(SwitchDateView.OnSwitchDateListener onSwitchDateListener) {
        mOnSwitchDateListener = onSwitchDateListener;
        return this;
    }

    public DailyAdapter setOnClickListener(View.OnClickListener onClickListener) {
        mOnClickListener = onClickListener;
        return this;
    }

    public DailyAdapter setOnRefreshCallback(OnRefreshCallback onRefreshCallback) {
        mOnRefreshCallback = onRefreshCallback;
        return this;
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }

    public void addItem(DailyReport dailyReport) {
        int size = mData.size();
        mData.add(dailyReport);
        notifyItemInserted(size);
    }

    public void initAddAll(List<DailyReport> dailyReports) {
        updateItem(0, dailyReports.get(dailyReports.size() - 1));
        insertDataToHead(dailyReports.subList(0, dailyReports.size() - 1));
    }

    public void addAll(List<DailyReport> dailyReports) {
        int size = mData.size();
        mData.addAll(dailyReports);
        notifyItemRangeInserted(size, dailyReports.size());
    }

    public void insertDataToHead(List<DailyReport> dailyReports) {
        mData.addAll(0, dailyReports);
        notifyItemRangeInserted(0, dailyReports.size());
    }

    public int updateItem(DailyReport dailyReport) {
        int position = getPosition(dailyReport.date);
        updateItem(position, dailyReport);
        return position;
    }

    public void updateItem(int position, DailyReport dailyReport) {
        if (position == -1) {
            return;
        }
        mData.set(position, dailyReport);
        notifyItemChanged(position);
    }

    public int getPosition(int unixTime) {
        for (int i = 0, length = mData.size(); i < length; i++) {
            DailyReport dailyReport = mData.get(i);
            if (dailyReport.date == unixTime) {
                return i;
            }
        }
        return -1;
    }

    public DailyReport getItem(int position) {
        return mData.get(position);
    }

    public long getItemDate(int position) {
        return mData.get(position).date;
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements SwipeRefreshLayout.OnRefreshListener {

        SwitchDateView mSwitchDateView;
        BlueRefreshView mBlueRefreshView;
        NestedScrollView mNestedScrollView;
        TouchDailySleepHistogramView mDaySleepHistogramView;
        ReportSleepDurationView mReportSleepDurationView;
        SleepNoteView mSleepNoteView;
        CardView mDoctorEvaluate;
        TextView mTvDoctorEvaluate;

        private DailyReport mDailyReport;

        private OnRefreshCallback mOnRefreshCallback;

        public ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);

            mSwitchDateView = itemView.findViewById(R.id.switch_date_view);
            mBlueRefreshView = itemView.findViewById(R.id.refresh);
            mNestedScrollView = itemView.findViewById(R.id.nested_scroll_view);
            mDaySleepHistogramView = itemView.findViewById(R.id.day_sleep_histogram_view);
            mReportSleepDurationView = itemView.findViewById(R.id.report_sleep_duration_view);
            mSleepNoteView = itemView.findViewById(R.id.sleep_note_view);
            mDoctorEvaluate = itemView.findViewById(R.id.doctor_evaluate);
            mTvDoctorEvaluate = itemView.findViewById(R.id.tv_doctor_evaluate);
        }

        public void initView(DailyReport dailyReport, SwitchDateView.OnSwitchDateListener switchDateListener, View.OnClickListener clickListener, OnRefreshCallback onRefreshCallback) {
            this.mDailyReport = dailyReport;
            this.mBlueRefreshView.setRefreshing(false);

            mOnRefreshCallback = onRefreshCallback;

            mSwitchDateView.setUnixTime(dailyReport.date);
            mSwitchDateView.setOnSwitchDateListener(switchDateListener);
            mBlueRefreshView.setOnRefreshListener(this);

            mDaySleepHistogramView.setData(dailyReport);
            mReportSleepDurationView.setSleepTodayDuration(mDailyReport.sleep_duration);
            mReportSleepDurationView.setLightSleepData(mDailyReport.light_duration, mDailyReport.light_duration_percent);
            mReportSleepDurationView.setDeepSleepData(mDailyReport.deep_duration, mDailyReport.deep_duration_percent);

            mSleepNoteView.addOnClickListener(clickListener);
            if (!TextUtils.isEmpty(mDailyReport.wrote_diary_at)) {

                SleepNote sleepNote = new SleepNote();
                sleepNote.wakeUpMood = mDailyReport.wake_up_mood;
                sleepNote.bedtimeState = mDailyReport.bedtime_state;
                sleepNote.remark = mDailyReport.remark;

                mSleepNoteView.setSleepNote(sleepNote);
                mSleepNoteView.show();
            } else {
                mSleepNoteView.hide();
            }

            if (!TextUtils.isEmpty(mDailyReport.doctors_evaluation)) {
                mDoctorEvaluate.setVisibility(View.VISIBLE);
                mTvDoctorEvaluate.setText(mDailyReport.doctors_evaluation);
            } else {
                mDoctorEvaluate.setVisibility(View.GONE);
            }

            if (mDailyReport.needScrollToBottom) {
                mNestedScrollView.post(() -> mNestedScrollView.fullScroll(ScrollView.FOCUS_DOWN));
                mDailyReport.needScrollToBottom = false;
            }
        }

        @Override
        public void onRefresh() {
            if (mOnRefreshCallback != null) {
                mOnRefreshCallback.onRefresh(getAdapterPosition(), mDailyReport);
            }
        }

        public void setRefreshing(boolean refreshing) {
            mBlueRefreshView.setRefreshing(refreshing);
        }
    }

    public interface OnRefreshCallback {

        void onRefresh(int position, DailyReport dailyReport);
    }
}

File: app/src/main/java/com/sumian/hw/report/adapter/DailyAdapter.kt
package com.sumian.hw.report.adapter

import android.support.v4.widget.NestedScrollView
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ScrollView
import com.sumian.hw.report.bean.DailyReport
import com.sumian.hw.report.widget.ReportSleepDurationView
import com.sumian.hw.report.widget.SwitchDateView
import com.sumian.hw.report.widget.histogram.TouchDailySleepHistogramView
import com.sumian.hw.widget.refresh.BlueRefreshView
import com.sumian.sd.R
import java.util.*

/**
 * Created by jzz
 * on 2018/3/12.
 * desc:
 */

class DailyAdapter : RecyclerView.Adapter<DailyAdapter.ViewHolder>() {

    private val mData: ArrayList<DailyReport>  by lazy {
        ArrayList<DailyReport>(0)
    }

    private lateinit var mOnSwitchDateListener: SwitchDateView.OnSwitchDateListener
    private var mOnRefreshCallback: OnRefreshCallback? = null


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val itemView = LayoutInflater.from(parent.context).inflate(R.layout.hw_lay_item_today_repot, parent, false)
        val viewHolder = ViewHolder(itemView)
        viewHolder.itemView.tag = viewHolder
        return viewHolder
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val dailyReport = mData[position]
        holder.initView(dailyReport, mOnSwitchDateListener, mOnRefreshCallback)
    }

    fun setOnSwitchDateListener(onSwitchDateListener: SwitchDateView.OnSwitchDateListener): DailyAdapter {
        mOnSwitchDateListener = onSwitchDateListener
        return this
    }

    fun setOnRefreshCallback(onRefreshCallback: OnRefreshCallback): DailyAdapter {
        mOnRefreshCallback = onRefreshCallback
        return this
    }

    override fun getItemCount(): Int {
        return mData.size
    }

    fun addItem(dailyReport: DailyReport) {
        val size = mData.size
        mData.add(dailyReport)
        notifyItemInserted(size)
    }

    fun initAddAll(dailyReports: List<DailyReport>) {
        updateItem(0, dailyReports[dailyReports.size - 1])
        insertDataToHead(dailyReports.subList(0, dailyReports.size - 1))
    }

    fun addAll(dailyReports: List<DailyReport>) {
        val size = mData.size
        mData.addAll(dailyReports)
        notifyItemRangeInserted(size, dailyReports.size)
    }

    fun insertDataToHead(dailyReports: List<DailyReport>) {
        mData.addAll(0, dailyReports)
        notifyItemRangeInserted(0, dailyReports.size)
    }

    fun updateItem(dailyReport: DailyReport): Int {
        val position = getPosition(dailyReport.date)
        updateItem(position, dailyReport)
        return position
    }

    fun updateItem(position: Int, dailyReport: DailyReport) {
        if (position == -1) {
            return
        }
        mData[position] = dailyReport
        notifyItemChanged(position)
    }

    fun getPosition(unixTime: Int): Int {
        var i = 0
        val length = mData.size
        while (i < length) {
            val dailyReport = mData[i]
            if (dailyReport.date == unixTime) {
                return i
            }
            i++
        }
        return -1
    }

    fun getItem(position: Int): DailyReport {
        return mData[position]
    }

    fun getItemDate(position: Int): Long {
        return mData[position].date.toLong()
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), SwipeRefreshLayout.OnRefreshListener {

        private val mSwitchDateView: SwitchDateView  by lazy {
            itemView.findViewById<SwitchDateView>(R.id.switch_date_view)
        }
        private val mBlueRefreshView: BlueRefreshView by lazy {
            itemView.findViewById<BlueRefreshView>(R.id.refresh)
        }
        private val mNestedScrollView: NestedScrollView  by lazy {
            itemView.findViewById<NestedScrollView>(R.id.nested_scroll_view)
        }
        private val mDaySleepHistogramView: TouchDailySleepHistogramView  by lazy {
            itemView.findViewById<TouchDailySleepHistogramView>(R.id.day_sleep_histogram_view)
        }
        private val mReportSleepDurationView: ReportSleepDurationView  by lazy {
            itemView.findViewById<ReportSleepDurationView>(R.id.report_sleep_duration_view)
        }

        private lateinit var mDailyReport: DailyReport

        private var mOnRefreshCallback: OnRefreshCallback? = null

        fun initView(dailyReport: DailyReport, switchDateListener: SwitchDateView.OnSwitchDateListener, onRefreshCallback: OnRefreshCallback?) {
            this.mDailyReport = dailyReport

            this.mBlueRefreshView.isRefreshing = false
            this.mBlueRefreshView.setOnRefreshListener(this)

            this.mOnRefreshCallback = onRefreshCallback

            this.mSwitchDateView.unixTime = dailyReport.date.toLong()
            this.mSwitchDateView.setOnSwitchDateListener(switchDateListener)

            this.mDaySleepHistogramView.setData(dailyReport)
            this.mReportSleepDurationView.setSleepTodayDuration(dailyReport.sleep_duration)
            this.mReportSleepDurationView.setLightSleepData(dailyReport.light_duration, dailyReport.light_duration_percent)
            this.mReportSleepDurationView.setDeepSleepData(dailyReport.deep_duration, dailyReport.deep_duration_percent)

            if (this.mDailyReport.needScrollToBottom) {
                this.mNestedScrollView.post { mNestedScrollView.fullScroll(ScrollView.FOCUS_DOWN) }
                this.mDailyReport.needScrollToBottom = false
            }
        }

        override fun onRefresh() {
            mOnRefreshCallback?.onRefresh(adapterPosition, mDailyReport)
        }

        fun setRefreshing(refreshing: Boolean) {
            mBlueRefreshView.isRefreshing = refreshing
        }
    }

    interface OnRefreshCallback {

        fun onRefresh(position: Int, dailyReport: DailyReport)
    }
}
